_source/_posts/2020-12-16-react-login.md

```diff
@@ -0,0 +1,599 @@
+Client ID: 0oa5qedkihI7QcSoi357
+```
+
+Modify the elements within the `BrowserRouter` component to match the code below.
+
+```jsx
+<BrowserRouter>
+  <Security oktaAuth={oktaAuth}>
+    <Header/>
+    <Route path='/' exact={true} component={Home}/>
+    <SecureRoute path='/private' exact={true} component={Private}/>
+    <Route path='/callback' component={LoginCallback}/>
+  </Security>
+</BrowserRouter>
+```
+
+You can see that the routes are now all contained within the `Security` component. This component makes the authentication service available to all its children. The `SecureRoute` component replaces `Route` for components that require authentication. You will also notice a new route on the `/callback` path. The `LoginCallback` component attached to this route handles the callback from the Okta servers once the user has logged on successfully.
+
+This is already enough to require authentication whenever the user wants to see the restricted page. To see how the authentication service can be used to create context-dependent content, open `src/Header.js`. Add the following import to the top of the file.
+
+```js
+import { useOktaAuth } from '@okta/okta-react';
+```
+
+Next, add code to create a sign-in or logout button depending on the current status. The `Header()` function should be modified to match the following.
+
+```jsx
+function Header() {
+ const { oktaAuth, authState } = useOktaAuth();
+
+  const login = () => { oktaAuth.signInWithRedirect(); }
+  const logout = () => { oktaAuth.signOut(); }
+
+  const userText = authState.isAuthenticated
+    ? <button onClick={ logout }>Logout</button>
+    : <button onClick={ login }>Sign In</button>;
+
+  return (
+    <header>
+      <div>React Login</div>
+      <ul className="menu">
+        <li><Link to="/">Home</Link></li>
+        <li><Link to="/private">Private</Link></li>
+      </ul>
+      {userText}
+    </header>
+  );
+}
+```
+
+Run `npm start` again to test your application. Now, when you click on **VISIT RESTRICTED PAGE**, you will be first taken to the sign-in page on your Okta domain. If you don't see a login form like the one below, it's likely because you're already logged in. Once you have signed in, you will be redirected to the protected page.
+
+{% img blog/react-login/react-login-okta-redirect.png alt:"The sign-in page hosted on the Okta servers" width:"800" %}{: .center-image }
+
+## Use the Okta Sign-In Widget for Login
+
+Redirecting the user to the Okta servers is appropriate in many situations. But sometimes you want to keep the user on your site without redirecting them to a server that they might not recognize. This is probably true for most public-facing web applications where you want to reach as many users as possible and reducing perceived barriers is important. The Okta Sign-In Widget allows you to embed the same user experience of the external login page into your website. To start, install the widget library using the following command.
```

Comment:
Replace "the user to the Okta servers" with "the user to an Okta hosted login page"

Reply:
Will do.